// @akashic/akashic-cli-commons/lib/ConsoleLogger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * ログを `console` に出力するロガー。
 */
var ConsoleLogger = /** @class */ (function () {
    function ConsoleLogger(param) {
        if (param === void 0) { param = {}; }
        this.quiet = !!param.quiet;
        this._log = param.debugLogMethod || console.log.bind(console);
    }
    /**
     * 任意の文字列を記録する。
     * @param message 記録するメッセージ
     */
    ConsoleLogger.prototype.print = function (message) {
        this._log(message);
    };
    /**
     * エラーを記録する。
     * @param message 記録するメッセージ
     * @param cause エラーの原因
     */
    ConsoleLogger.prototype.error = function (message, cause) {
        this.print("ERROR: " + message);
        if (cause)
            this.print(cause);
    };
    /**
     * 警告を記録する。
     * @param message 記録するメッセージ
     * @param cause 警告の原因
     */
    ConsoleLogger.prototype.warn = function (message, cause) {
        this.print("WARN: " + message);
        if (cause)
            this.print(cause);
    };
    /**
     * 情報を記録する。
     * @param message 記録するメッセージ
     * @param cause 情報の原因
     */
    ConsoleLogger.prototype.info = function (message, cause) {
        if (this.quiet)
            return;
        this.print("INFO: " + message);
        if (cause)
            this.print(cause);
    };
    return ConsoleLogger;
}());
exports.ConsoleLogger = ConsoleLogger;
